### PR TITLE
fix(ext/http): close stream on resp body error

### PR DESCRIPTION
**Pull request by lucacasonato**
Previously, errored streaming response bodies did not cause the HTTP stream
to be aborted. It instead caused the stream to be closed gracefully,
which had the result that the client could not detect the difference
between a successful response and an errored response.

This commit fixes the issue by aborting the stream on error.

### CODE REVIEW

File: ext/http/reader_stream.rs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    writer.write_all(b"hello").await.unwrap();
+    assert_eq!(stream.next().await.unwrap().unwrap(), Bytes::from("hello"));
+
+    writer.write_all(b"world").await.unwrap();
+    assert_eq!(stream.next().await.unwrap().unwrap(), Bytes::from("world"));
+
+    shutdown_handle.shutdown();
+    writer.shutdown().await.unwrap();
+    drop(writer);
+    assert!(stream.next().await.is_none());
```
Comment (ry) [inline ext/http/reader_stream.rs:92]: what happens if you insert another `writer.write_all(b"foo");` either before or after the shutdown?

File: ext/http/reader_stream.rs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    self: Pin<&mut Self>,
+    cx: &mut Context<'_>,
+  ) -> Poll<Option<Self::Item>> {
+    let this = self.project();
+    let val = std::task::ready!(this.inner.poll_next(cx));
+    match val {
+      None if this.done.load(Ordering::Relaxed) => Poll::Ready(None),
+      None => Poll::Ready(Some(Err(std::io::Error::new(
+        std::io::ErrorKind::UnexpectedEof,
+        "channel closed",
```
[the remark this answers is not in the document]
Comment (lucacasonato) [reply ext/http/reader_stream.rs:60]: This error is never exposed to users - it is swallowed inside of our `hyper` integration. But I'll make it more specific.

File: ext/http/reader_stream.rs (hunk 1 above)
Comment (lucacasonato) [reply ext/http/reader_stream.rs:92]: calling `poll_write` after `poll_shutdown` results in a `std::io::ErrorKind::BrokenPipe`

Comment (lucacasonato) [reply ext/http/reader_stream.rs:92]: before the shutdown it will just complete (this example does two writes before the shutdown)

Comment (ry) [reply ext/http/reader_stream.rs:92]: Maybe you can these edge cases either in this test or another one?

```rust
writer.write_all(b"foo").unwrap();
shutdown_handle.shutdown();
assert!(stream.next().await.is_none()); // correct?
writer.shutdown().await.unwrap();
assert_eq!(writer.write_all(b"foo").unwrap_err(), "broken pipe");
drop(writer);
assert!(stream.next().await.is_none());
```

File: cli/tests/unit/http_test.ts [hunk 3]
```diff
@@ -2614,6 +2614,129 @@ Deno.test({
...
+async function httpServerWithErrorBody(
+  listener: Deno.Listener,
+  compression: boolean,
+): Promise<Deno.HttpConn> {
+  const conn = await listener.accept();
+  listener.close();
+  const httpConn = Deno.serveHttp(conn);
+  const e = await httpConn.nextRequest();
+  assert(e);
+  const { respondWith } = e;
```
Comment (ry) [inline cli/tests/unit/http_test.ts:2626]: maybe you should assert something here about the request being H1 or H2

File: ext/http/reader_stream.rs (hunk 1 above)
Comment (lucacasonato) [reply ext/http/reader_stream.rs:92]: except for that `// correct?`, yes - that line will hang until `writer()` is dropped

File: cli/tests/unit/http_test.ts (hunk 3 above)
Comment (lucacasonato) [reply cli/tests/unit/http_test.ts:2626]: Let me see if I can find a way to distinguish them. Unfortunately we do not expose the HTTP version on the `Request` (it's a property we should add to `Deno.HttpConn` I think)

File: ext/http/lib.rs [hunk 4]
```diff
@@ -425,6 +428,39 @@ impl Default for HttpResponseWriter {
...
 }
 
+struct BodyCompressedHolder {
+  pub writer: Pin<Box<dyn tokio::io::AsyncWrite>>,
+  pub shutdown_handle: ShutdownHandle,
+}
+
+struct BodyUncompressedSender {
+  sender: Option<hyper::body::Sender>,
+  done: bool,
```
Comment (piscisaureus) [inline ext/http/lib.rs:438]: What's the reason for having this as a separate `bool`, rather than just `.take()`ing the stream once done?

File: ext/http/lib.rs [hunk 5]
```diff
@@ -425,6 +428,39 @@ impl Default for HttpResponseWriter {
   }
 }
 
+struct BodyCompressedHolder {
+  pub writer: Pin<Box<dyn tokio::io::AsyncWrite>>,
+  pub shutdown_handle: ShutdownHandle,
+}
+
+struct BodyUncompressedSender {
+  sender: Option<hyper::body::Sender>,
```
Comment (piscisaureus) [inline ext/http/lib.rs:437]: Answer the question below first. But if you decide that you definitely want a boolean, drop the `Option` wrapper.

File: ext/http/reader_stream.rs (hunk 1 above)
Comment (piscisaureus) [reply ext/http/reader_stream.rs:92]: > calling poll_write after poll_shutdown results in a std::io::ErrorKind::BrokenPipe

That sounds like the correct behavior to me.

File: ext/http/reader_stream.rs [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+  #[pin]
+  inner: ReaderStream<R>,
+  done: Arc<AtomicBool>,
+}
+
+pub struct ShutdownHandle(Arc<AtomicBool>);
+
+impl ShutdownHandle {
+  pub fn shutdown(&self) {
+    self.0.store(true, std::sync::atomic::Ordering::SeqCst);
```
Comment (piscisaureus) [inline ext/http/reader_stream.rs:32]: Where does this ever get read? 
It looks to me that this whole ShutdownHandle thing is unnecessary.

File: ext/http/reader_stream.rs [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+  type Item = std::io::Result<Bytes>;
+
+  fn poll_next(
+    self: Pin<&mut Self>,
+    cx: &mut Context<'_>,
+  ) -> Poll<Option<Self::Item>> {
+    let this = self.project();
+    let val = std::task::ready!(this.inner.poll_next(cx));
+    match val {
+      None if this.done.load(Ordering::Relaxed) => Poll::Ready(None),
```
Comment (piscisaureus) [inline ext/http/reader_stream.rs:57]: I think you need this load to be SeqCst too.

Comment (piscisaureus) [reply ext/http/reader_stream.rs:57]: Also you might not get to this point even if you expect to.
If no more data is sent on `inner` you might never get past  this line above: `let val = std::task::ready!(this.inner.poll_next(cx));`, if only because the waker that would schedule the current task never gets woken.

File: ext/http/lib.rs [hunk 8]
```diff
@@ -425,6 +428,39 @@ impl Default for HttpResponseWriter {
   }
 }
 
+struct BodyCompressedHolder {
```
Comment (piscisaureus) [inline ext/http/lib.rs:431]: It would be a little cleaner if `BodyCompressedHolder` itself just implemented AsyncWrite.

File: cli/tests/unit/http_test.ts (hunk 3 above)
Comment (piscisaureus) [reply cli/tests/unit/http_test.ts:2626]: > (it's a property we should add to Deno.HttpConn I think)

That makes sense, but for std/http users that is no good because the conn is not exposed.

Comment (piscisaureus) [reply cli/tests/unit/http_test.ts:2626]: Seems more like a WinterCG project to add it to the Request spec?

File: ext/http/lib.rs (hunk 8 above)
Comment (piscisaureus) [reply ext/http/lib.rs:431]: If you don't want to do that and make this essentially a POD struct, why not define:

```rust
enum HttpResponseWriter {
  // ...
  Body { writer: Pin<Box<dyn tokio::io::AsyncWrite>>, is_shut: Arc<AtomicBool> }
  // ...
}
```

File: ext/http/reader_stream.rs (hunk 7 above)
Comment (lucacasonato) [reply ext/http/reader_stream.rs:57]: Yes, which is why in the only path that sets `this.done` also immediately calls shutdown on the writable end of the stream, triggering `this.inner.poll_next` to resolve with `Poll::Ready(None)`.

File: ext/http/reader_stream.rs (hunk 1 above)
Comment (lucacasonato) [reply ext/http/reader_stream.rs:92]: I agree, that is the correct behaviour. I can add a test that calls write after shutdown though for prosperity.

File: cli/tests/unit/http_test.ts (hunk 3 above)
Comment (lucacasonato) [reply cli/tests/unit/http_test.ts:2626]: > That makes sense, but for std/http users that is no good because the conn is not exposed.

Things on `Conn` are exposed via the second argument to `serve()` (the `ConnInfo` struct).

File: ext/http/lib.rs (hunk 4 above)
Comment (lucacasonato) [reply ext/http/lib.rs]: Yes, good point. I restructured the code to get rid of it.

File: cli/tests/unit/http_test.ts (hunk 3 above)
Comment (piscisaureus) [reply cli/tests/unit/http_test.ts:2626]: > > That makes sense, but for std/http users that is no good because the conn is not exposed.
> 
> Things on `Conn` are exposed via the second argument to `serve()` (the `ConnInfo` struct).

I guess that could work.
Still would be nice to be on Request though, especially if the property was also available to set the HTTP version on outgoing requests.

**Review by piscisaureus — APPROVED** [review]
